x-pack/plugins/drilldowns/public/types.ts

```diff
@@ -0,0 +1,127 @@
+   * helper from `kibana_utils` plugin.
+   *
+   * ```tsx
+   * import React from 'react';
+   * import { uiToReactComponent } from 'src/plugins/kibana_utils';
+   * import { UiActionsCollectConfigProps as CollectConfigProps } from 'src/plugins/ui_actions/public';
+   *
+   * type Props = CollectConfigProps<Config>;
+   *
+   * const ReactCollectConfig: React.FC<Props> = () => {
+   *   return <div>Collecting config...'</div>;
+   * };
+   *
+   * export const CollectConfig = uiToReactComponent(ReactCollectConfig);
+   * ```
+   */
+  CollectConfig: ActionFactoryDefinition<
+    Config,
+    DrilldownFactoryContext<CreationContext>,
+    ExecutionContext
+  >['CollectConfig'];
+
+  /**
+   * A validator function for the config object. Should always return a boolean
+   * given any input.
+   */
+  isConfigValid: ActionFactoryDefinition<
+    Config,
+    DrilldownFactoryContext<CreationContext>,
+    ExecutionContext
+  >['isConfigValid'];
+
+  /**
+   * Name of EUI icon to display when showing this drilldown to user.
+   */
+  euiIcon?: string;
+
+  /**
+   * Should return an internationalized name of the drilldown, which will be
+   * displayed to the user.
+   */
+  getDisplayName: () => string;
+
+  /**
+   * Whether this drilldown should be considered for execution given `config`
+   * and `context`. When multiple drilldowns are attached to the same trigger
+   * user is presented with a context menu to pick one drilldown for execute. If
+   * this method returns `true` this trigger will appear in the context menu
+   * list, if `false`, it will not be presented to the user. If `doExecute` is
+   * not implemented, this drilldown will always be show to the user.
+   *
+   * @param config Config object that user configured this drilldown with.
+   * @param context Object that represents context in which the underlying
+   *  `UIAction` of this drilldown is being executed in.
+   */
+  doExecute?(config: Config, context: ExecutionContext): Promise<boolean>;
+
+  /**
+   * Implements the "navigation" action of the drilldown. This happens when
+   * user clicks something in the UI that executes a trigger to which this
+   * drilldown was attached.
+   *
+   * @param config Config object that user configured this drilldown with.
+   * @param context Object that represents context in which the underlying
+   *  `UIAction` of this drilldown is being executed in.
+   */
+  execute(config: Config, context: ExecutionContext): void;
+}
+
+/**
+ * Context object used when creating a drilldown.
+ */
+export interface DrilldownFactoryContext<T> {
+  /**
+   * Place where factory is being rendered.
+   */
+  place?: string;
+
+  /**
+   * Context provided to the drilldown factory by the place where the UI is
+   * rendered. For example, for the "dashboard" place, this context contains
+   * the ID of the current dashboard, which could be used for filtering it out
+   * of the list.
+   */
+  placeContext: T;
```

Comment:
`placeContext` is not used for "only show in certain app" it is a bit different. It allows the caller of drilldown flyout to pass in some context information about the place from where the drilldown flyout was opened. In Dashboard case `placeContext` is the embeddable context. For example, it can be used for action factories to infer the current dashboard so they don't display it in the picker.

Matt's Dashboard-to-Dashboard PR it is currently in progress, but this is how it would be used:

![image](https://user-images.githubusercontent.com/9773803/77151100-96268680-6a95-11ea-8dd7-e6644ec988b4.png)